### DISCUSSION TITLE
How to acknowledge zero length bulk IN transfers

### DISCUSSION THREAD

**Discussion by thierer** [hathach/tinyusb#903] [Q&A]
What is the correct way to acknowledge a bulk IN request with a zero length response (to signal that no data is available)?

Right now, this seems to be prevented by `vendor_device.c::maybe_transmit()`, which only starts the transfer if there is data in the FIFO (`count > 0`): https://github.com/hathach/tinyusb/blob/81fa9966dedaeae23d0eb435b2377ae4695151c5/src/class/vendor/vendor_device.c#L109-L120

I came up with https://github.com/thierer/tinyusb/commit/dfb2ecc6891d47f500097e93abe559343dafbd97, which seems to do what I want, but maybe there's a better way to achieve this - without modifying tinusb or reimplementing most of `vendor_device`?

**Comment by kkitayam** [comment]
My understanding that it is implemented with a FIFO. This means that the application does not need to be aware of packet-level transfers. In other words, an application cannot handle packet-level transfers.
If you want to handle packet-level transfers, you can implement a vendor class from scratch. It was mentioned in #866 (review).

**Reply by hathach** [reply]
hmm, currently not yet, unless like @kkitayam said to implement your own driver. However, I think we could expand the API to allow `tud_vendor_write(NULL,0)` to send ZLP when there is no data in the fifo. If you are willing to submit the PR, I will be happy to review it.

**Reply by thierer** [reply]
@kkitayam I might end up doing that. Thanks for the pointer to the PR, the examples mentioned there might come handy.

@hathach Sure. Did you take a look at the commit I linked to? From what I understand you'd prefer to make the transfer depend on buffer == NULL (in addition to length == 0)?

**Reply by hathach** [reply]
Can you submit it as draft PR. It makes it easier to discuss code change there. Ultimate I don't to change the behavior of maybe_transmit(), the idea is write(NULL,0) when calling with empty fifo will trigger an zlp packet.
